piplicenses/__init__.py

```diff
@@ -112,16 +114,71 @@
 }
 
 
-SYSTEM_PACKAGES = (
+SYSTEM_PACKAGES = [
     __pkgname__,
     "pip",
     "pip-licenses-lib",
     "prettytable",
     "wcwidth",
     "setuptools",
-    "tomli",
     "wheel",
-)
+]
+if sys.version_info < (3, 11):
+    SYSTEM_PACKAGES.append("tomli")
+
+
+class SpdxParser(Protocol):
+    def __call__(self, expression: str) -> set[str]: ...
+
+
+@functools.lru_cache(maxsize=1)
+def _get_spdx_parser() -> SpdxParser:
+    """
+    Create an SPDX expression parser.
+
+    If the extra "spdx" is not installed, then the parser just returns a set
+    with the provided expression as only element.
+    """
+    try:
+        from license_expression import get_spdx_licensing
+
+    except ImportError:
+
+        def dummy_parser(expression: str) -> set[str]:
+            return {expression}
+
+        return dummy_parser
+
+    SYSTEM_PACKAGES.append("license-expression")
+    SYSTEM_PACKAGES.append("boolean-py")
+
+    licensing = get_spdx_licensing()
+
+    def parser(expression: str) -> set[str]:
+        try:
+            result = licensing.validate(expression)
+        except Exception as exception:
+            # https://github.com/aboutcode-org/license-expression/issues/97
+            return {expression}
+        if result.errors:
+            return {expression}
+        parsed = licensing.parse(expression)
+        if parsed is None:
+            return {expression}
+        parsed = parsed.simplify()
+        if "AND" in str(parsed) or "WITH" in str(parsed):
+            warnings.warn(
```

Comment:
Warnings without a stacklevel tend to be discouraged. Additionally, it usually is nice to have a dedicated warning class instead of the default *UserWarning*.

Apart from this implementation detail: Could you please add at least a simple unit test for this case as well?